src/gretel_trainer/relational/json.py

```diff
@@ -367,16 +390,32 @@ def _denormalize_json(
                 col_name = get_parent_column_name_from_child_table_name(
                     table_provenance_name
                 )
-                table_dict[parent_name] = table_dict[parent_name].join(
-                    nested_df.rename(col_name)
-                )
-                table_dict[parent_name][col_name] = nulls_to_empty_lists(
-                    table_dict[parent_name][col_name]
-                )
+                parent_df = get_table(parent_node_name)
+                parent_df = parent_df.join(nested_df.rename(col_name))
+                parent_df[col_name] = nulls_to_empty_lists(parent_df[col_name])
+                table_dict[parent_provenance_name] = parent_df
             table_dict[table_provenance_name] = nested_df
     return table_dict[original_table_name]
 
 
+def _get_table(
+    rel_data: _RelationalData,
+    inverse_table_name_mappings: dict[str, str],
+    table_dict: dict[str, pd.DataFrame],
+    node_name: str,
+) -> pd.DataFrame:
+    """
+    Helper function used inside _denormalize_json to safely retrieve a table dataframe.
+    If `node_name` is not present in `table_dict` (e.g. because the Gretel Job failed),
+    return an empty dataframe with the expected columns.
+    """
+    empty_fallback = pd.DataFrame(
+        data={col: [] for col in rel_data.get_table_columns(node_name)},
```

Comment:
nit, could also do `pd.DataFrame(columns=rel_data.get_table_columns(node_name))`
```
Pd.>>> pd.DataFrame(columns=["a","b","c"])
Empty DataFrame
Columns: [a, b, c]
Index: []
>>> pd.DataFrame(data={col:[] for col in ["a","b","c"]})
Empty DataFrame
Columns: [a, b, c]
Index: []
```

Reply:
Yep, yours is definitely better